lib/tasks/upgrade.rake

```diff
@@ -0,0 +1,56 @@
+# TRANSLATORS: do not translate
+desc <<-END_DESC
+  This task runs predefined upgrade steps.
+
+  Examples:
+    rake upgrade:run
+    rake upgrade:mark_as_ran TASK_NAME=some_rake_task,another_rake_task
+
+END_DESC
+
+namespace :upgrade do
+  task :run => :environment do
+    ENV['FOREMAN_UPGRADE'] = '1'
+    tasks = Foreman::Plugin.registered_plugins.values.map{|plugin| plugin.upgrade_tasks}.flatten
+    tasks.select!{|task| task.run?}
+
+    tasks.each_with_index do |task, index|
+      count = "#{index + 1}/#{tasks.count}"
+
+      message = if task.long_running?
+                  _("Upgrade Step %{count}: %{name}, This may take a long while.")
+                else
+                  _("Upgrade Step %{count}: %{name}")
+                end
+
+      puts '============================================='
+      puts message % {:name => task.name, :count => count}
+      task.mark_as_ran if run_task(task)
```

Comment:
So the assumption is:

 1) If it runs successfully, a database object will exist that represents the task
 2) If an upgrade task is ran, but fails, it will not be present in the database

So a lack of a database object to represent a task means either it never ran or it failed during the run. I question whether we should be tracking the latter explicitly for debugging purposes.

Reply:
what's the value in that exactly?  Logs i think would give us wayyyy more information.  When would it be useful to know that a task ran and failed, but not have any information as to why?

Reply:
Or maybe introduce an upgrade logger?